### PR TITLE
Getting Started Guide for Windows Support

### PR DESCRIPTION
**Pull request by jbhurat**
This PR is to add documentation for getting started guide for supporting Windows Server Containers on Kubernetes.
The work to support Windows Server Container is tracked by kubernetes/features#116 and the code changes are already merged and tagged for Release 1.5

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1771)

### CODE REVIEW

**Comment by errordeveloper** [comment]
Have you tried rendering this? I mostly wonder if it will get added to the site navigation automatically, or you need to do something special? Also, I have very little idea about anything Windows-specific, so I think it needs a different assignee...

**Comment by errordeveloper** [comment]
ping @jaredbhatti

**Comment by jbhurat** [comment]
@errordeveloper I am not sure if it will get added to the site navigation automatically. @jaredbhatti might be able to help here

File: docs/getting-started-guides/windows/README.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+# Running Windows Server Containers using Kubernetes
+Microsoft, in collaboration with Docker, is introducing a new feature called Windows Server Containers with Windows Server 2016, which enables Docker containers to run on Windows. We have enhanced Kubernetes to support Windows Server Containers (and Windows Server 2016 as a container host node).
```
Comment (erictune) [inline docs/getting-started-guides/windows/README.md]: Kubernetes documentation typically does not highlight the work of a specific company, group or individual.  (A blog post is a great place for giving credit.)  

With that in mind, I'd suggest a couple of rewordings:

The first sentence could be reworded as:

I'd suggest replacing "we" everywhere in this doc.  For example,   just say "Kubernetes supports Windows Server Containers ...." 

I'd also suggest rewording the first two sentences to  something like:

"Kubernetes supports nodes running Microsoft Windows Server 2016.  Docker version x.xx is required .  Pods are comprised of Windows Server Containers."  

You can definitely link from various terms (e.g. Windows Server Containers) to product pages on the Microsoft and Docker websites.

File: docs/getting-started-guides/windows/README.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+4. Docker Version 1.12.2-cs2-ws-beta or later
+
+## Networking
+Network is achieved using L3 routing. Because third-party networking plugins (e.g. flannel, calico, etc) don’t natively work on Windows Server, we relied on existing technology that is built into the Windows and Linux operating systems. In this L3 networking approach, we chose a /16 subnet for the cluster nodes, and we assign a /24 subnet to each worker node. All pods on a given worker node will be connected to the /24 subnet. This allows pods on the same node to communicate with each other. In order to enable networking between pods running on different nodes, routing features that are built into Windows Server 2016 and Linux are used.
+
+### Linux
+The above networking approach is already supported on Linux using a bridge interface, which essentially creates a private network local to the node. Similar to the Windows side, routes to all other pod CIDRs must be created in order to send packets via the “public” NIC.
+
+### Windows
+To support networking requirements, we will need the following configuration on each Kubernetes Windows Server node:
```
Comment (erictune) [inline docs/getting-started-guides/windows/README.md]: Replace "To support networking requirements, we will need the following configuration on each Kubernetes Windows Server node" with "Each Window Server node should have the following configuration".

File: docs/getting-started-guides/windows/README.md (hunk 1 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Updated as requested

File: docs/getting-started-guides/windows/README.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
+# Running Windows Server Containers using Kubernetes
+Microsoft, in collaboration with Docker, is introducing a new feature called Windows Server Containers with Windows Server 2016, which enables Docker containers to run on Windows. We have enhanced Kubernetes to support Windows Server Containers (and Windows Server 2016 as a container host node).
+
+Windows Server and .NET still account for a significant portion of the workloads in the enterprise, and with this work, customers will be able to run Windows-based and .NET-based applications inside containers on Kubernetes, making Kubernetes the first fully functional cross-platform cluster manager. Services that span Windows-based and Linux-based containers are now a reality.
+
+Windows Server Containers will be supported on Kubernetes as an alpha feature. Kubernetes control plane (API Server, Scheduler, Controller Manager etc) continue to run on Linux, while the kubelet and kube-proxy can be run on Windows Server.
```
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Updated as requested

File: docs/getting-started-guides/windows/README.md (hunk 2 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Updated as requested

**Comment by jbhurat** [comment]
Thanks for the review @erictune . I have updated the documentation as requested

**Comment by jbhurat** [comment]
@erictune Can you please review the changes and let me know if your concerns are addressed

File: docs/getting-started-guides/windows/README.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
+# Running Windows Server Containers using Kubernetes
+Windows Server Containers are supported on Kubernetes as an alpha feature. Kubernetes control plane (API Server, Scheduler, Controller Manager etc) will continue to run on Linux, while the kubelet and kube-proxy can be run on Windows Server.
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Reference a specific version; it helps the documentation from being rendered obsolete.

"Kubernetes version 1.5 introduces support for Windows Server Containers. In version 1.5, the Kubernetes control plane (API Server, Scheduler, Controller Manager, etc) continue to run on Linux, while the kubelet and kube-proxy can be run on Windows Server.

**Note:** Windows Server Containers on Kubernetes is an Alpha feature in Kubernetes 1.5."

File: docs/getting-started-guides/windows/README.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+## Prerequisites
+With the alpha release, Windows Server Containers for Kubernetes is supported using the following:
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: "With the alpha release," -> "In Kubernetes version 1.5,"

File: docs/getting-started-guides/windows/README.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+1. Two NICs (virtual networking adapters) are required on each Windows Server node - The two Windows container networking modes of interest (transparent and L2 bridge) use an external Hyper-V virtual switch. This means that one of the NICs is entirely allocated to the bridge, creating the need for the second NIC.
+2. Transparent container network created - This is a manual configuration step and is shown in **_Route Setup_** section below
+3. RRAS (Routing) Windows feature enabled - Allows routing between NICs on the box, and also “captures” packets that have the destination IP of a POD running on the node. To enable, open “Server Manager”. Click on “Roles”, “Add Roles”. Click “Next”. Select “Network Policy and Access Services”. Click on “Routing and Remote Access Service” and the underlying checkboxes
+4. Routes defined pointing to the other pod CIDRs via the “public” NIC - These routes are added to the built-in routing table as shown in **_Route Setup_** section below
+
+The following diagram illustrates the Windows Server networking setup for Kubernetes Setup
+![Windows Setup](windows-setup.png)
+
+## Setup
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Three successive nested headings is indicative of major structural problems.

Better structure:

## Setting up Windows Server Containers on Kubernetes

To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows. The setup requirements for Windows Server Containers vary depending on whether or not you're using Windows or Linux Hosts.

### Windows Host Setup

1. Steps

### Linux Host Setup (editorial note: Why? I thought the whole point of this document was how to do Windows node setups).

1. Steps

### Component Setup

Requirements

File: docs/getting-started-guides/windows/README.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+### Component Setup
+In order to build the work node components, the *kubelet* and *kube-proxy* for Windows, the following needs to be installed on the host
+* Git, Go 1.7.1+ 
+* make (if using Linux or MacOS)
+* Important notes and other dependencies are listed [here](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#building-kubernetes-on-a-local-osshell-environment)
+
+#### kubelet
+
+In order to build the *kubelet*, run:
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: "In order" not necessary.

File: docs/getting-started-guides/windows/README.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+2. DNS support for Windows recently got merged to docker master and is currently not supported in a stable docker release. To use DNS build docker from master or download the binary from [Docker master](https://master.dockerproject.org/)
+3. Pull the `apprenda/pause` image from `https://hub.docker.com/r/apprenda/pause` 
+4. RRAS (Routing) Windows feature enabled
+
+#### Linux
+
+Linux hosts should be setup according to their respective distro documentation and the requirements of the Kubernetes version you will be using. Linux hosts also require to have CNI Setup.
+
+### Component Setup
+In order to build the work node components, the *kubelet* and *kube-proxy* for Windows, the following needs to be installed on the host
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: "In order" not necessary.

File: docs/getting-started-guides/windows/README.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+In order to build the *kubelet*, run:
+
+1. `cd $GOPATH/src/k8s.io/kubernetes`
+2. Build *kubelet*
+   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kubelet`
+   2. Windows: `go build cmd/kubelet/kubelet.go`
+
+#### kube-proxy
+
+In order to build *kube-proxy*, run:
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: "In order" not necessary.

File: docs/getting-started-guides/windows/README.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+```
+docker network create -d transparent --gateway 192.168.2.1 --subnet 192.168.2.0/24 <network name>
+# A bridge is created with Adapter name "vEthernet (HNSTransparent)". Set its IP address to transparent network gateway
+netsh interface ipv4 set address "vEthernet (HNSTransparent)" addr=192.168.2.1
+route add 192.168.0.0 mask 255.255.255.0 192.168.0.1 if <Interface Id of the Routable Ethernet Adapter> -p
+route add 192.168.1.0 mask 255.255.255.0 192.168.1.1 if <Interface Id of the Routable Ethernet Adapter> -p
+```
+
+## Starting the Cluster
+For now, the Kubernetes control plane continues to run on Linux and as a result a Windows only Kubernetes Cluster is not possible.
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Why this sentence? We covered this in the introduction.

Suggestion: "To start your cluster, you'll need to start both the Linux-based Kubernetes control plane, and the Windows Server-based Kubernetes node components."

File: docs/getting-started-guides/windows/README.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+docker network create -d transparent --gateway 192.168.2.1 --subnet 192.168.2.0/24 <network name>
+# A bridge is created with Adapter name "vEthernet (HNSTransparent)". Set its IP address to transparent network gateway
+netsh interface ipv4 set address "vEthernet (HNSTransparent)" addr=192.168.2.1
+route add 192.168.0.0 mask 255.255.255.0 192.168.0.1 if <Interface Id of the Routable Ethernet Adapter> -p
+route add 192.168.1.0 mask 255.255.255.0 192.168.1.1 if <Interface Id of the Routable Ethernet Adapter> -p
+```
+
+## Starting the Cluster
+For now, the Kubernetes control plane continues to run on Linux and as a result a Windows only Kubernetes Cluster is not possible. 
+## Linux
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: ## Starting the Linux-based Control Plane

File: docs/getting-started-guides/windows/README.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+Use your preferred method to start Kubernetes cluster on Linux. Please note that Cluster CIDR might need to be updated.
+## Windows
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: ## Starting the Windows Node Components

File: docs/getting-started-guides/windows/README.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+### kubelet
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Nested headings. Replace "### kubelet" with "To start kubelet on your Windows node:"

File: docs/getting-started-guides/windows/README.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+Run the following in a PowerShell window. Be aware that if the node reboots or the process exits, you will have to rerun the commands below to restart the kubelet
+
+1. Set environment variable *CONTAINER_NETWORK* value to the docker container network to use
+`$env:CONTAINER_NETWORK = "<docker network>"`
+
+2. Run *kubelet* executable using the below command
+`kubelet.exe --hostname-override=<ip address/hostname of the windows node>  --pod-infra-container-image="apprenda/pause" --resolv-conf="" --api_servers=<api server location>`
+
+### kube-proxy
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Replace "### kube-proxy" with "To start kube-proxy on your Windows node:"

File: docs/getting-started-guides/windows/README.md [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+Run the following in a PowerShell window with administrative privileges. Be aware that if the node reboots or the process exits, you will have to rerun the commands below to restart the kube-proxy.
+
+1. Set environment variable *INTERFACE_TO_ADD_SERVICE_IP* value to a node only network interface. The interface created when docker is installed should work
+`$env:INTERFACE_TO_ADD_SERVICE_IP = "vEthernet (HNS Internal NIC)"`
+
+2. Run *kube-proxy* executable using the below command
+`.\proxy.exe --v=3 --proxy-mode=userspace --hostname-override=<ip address/hostname of the windows node> --master=<api server location> --bind-address=<ip address of the windows node>`
+
+## Scheduling PODs on Windows
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: "Pod" should not be all-caps. It's not an acronym.

File: docs/getting-started-guides/windows/README.md [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+As Kubernetes control plane currently runs on Linux, the resulting cluster will have both Linux and Windows nodes. TO schedule PODs on Windows, `nodeSelector` constraint has to be set with label `beta.kubernetes.io/os` having value `windows` as shown in the example below,
```
Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Neither "TO" nor "POD" should be all-caps.

Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Suggested rephrase for clarity: "Because your cluster has both Linux and Windows nodes, you must explictly set the `nodeSelector` constraint to be able to schedule Pods to Windows nodes. You must set `nodeSelector` with the label `beta.kubernetes.io/os` to the value `windows`; see the following example:

File: docs/getting-started-guides/windows/README.md (hunk 6 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: The reason I have 3 successive nested headings is because Setup have 3 types Host, Component and Route. Host and Route instructions are different for Windows and Linux Hosts and Components setup have instructions for building `kubelet` and `kube-proxy` for Windows. So as each Setup type have 2 more child elements I thought it is easier to follow if it is nested headings.

File: docs/getting-started-guides/windows/README.md (hunk 4 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Thanks for the review @devin-donnelly. I am making the suggested changes

File: docs/getting-started-guides/windows/README.md (hunk 5 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Thanks for the review @devin-donnelly. I am making the suggested changes

File: docs/getting-started-guides/windows/README.md (hunk 8 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Thanks for the review @devin-donnelly. I am making the suggested changes

File: docs/getting-started-guides/windows/README.md (hunk 7 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Thanks for the review @devin-donnelly. I am making the suggested changes

File: docs/getting-started-guides/windows/README.md (hunk 9 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestion

File: docs/getting-started-guides/windows/README.md (hunk 10 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions

File: docs/getting-started-guides/windows/README.md (hunk 11 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 12 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 13 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 14 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 15 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 16 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: I am accepting the suggestions and making the required changes

File: docs/getting-started-guides/windows/README.md (hunk 6 above)
Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: ## Setting up Windows Server Containers on Kubernetes
To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows and setup Routes for Pod communication on different nodes
### Host Setup
**Windows Host Setup**
1. Windows Server container host running Windows Server 2016 and Docker v1.12. Follow the setup instructions outlined by this blog post: https://msdn.microsoft.com/en-us/virtualization/windowscontainers/quick_start/quick_start_windows_server
2. DNS support for Windows recently got merged to docker master and is currently not supported in a stable docker release. To use DNS build docker from master or download the binary from [Docker master](https://master.dockerproject.org/)
3. Pull the `apprenda/pause` image from `https://hub.docker.com/r/apprenda/pause` 
4. RRAS (Routing) Windows feature enabled

**Linux Host Setup**
1. Linux hosts should be setup according to their respective distro documentation and the requirements of the Kubernetes version you will be using. 
2. CNI network plugin installed.

### Component Setup
Requirements
* Git, Go 1.7.1+ 
* make (if using Linux or MacOS)
* Important notes and other dependencies are listed [here](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#building-kubernetes-on-a-local-osshell-environment)

**kubelet**
To build the *kubelet*, run:

1. `cd $GOPATH/src/k8s.io/kubernetes`
2. Build *kubelet*
   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kubelet`
   2. Windows: `go build cmd/kubelet/kubelet.go`

**kube-proxy**
To build *kube-proxy*, run:

1. `cd $GOPATH/src/k8s.io/kubernetes`
2. Build *kube-proxy*
   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kube-proxy`
   2. Windows: `go build cmd/kube-proxy/proxy.go`

### Route Setup
The below example setup assumes one Linux and two Windows Server 2016 nodes and a cluster CIDR 192.168.0.0/16

| Hostname | Routable IP address | Pod CIDR |
| --- | --- | --- |
| Lin01 | `<IP of Lin01 host>` | 192.168.0.0/24 |
| Win01 | `<IP of Win01 host>` | 192.168.1.0/24 |
| Win02 | `<IP of Win02 host>` | 192.168.2.0/24 |

**Lin01**
```
ip route add 192.168.1.0/24 via <IP of Win01 host>
ip route add 192.168.2.0/24 via <IP of Win02 host>
```

Comment (jbhurat) [inline docs/getting-started-guides/windows/README.md]: Is this acceptable resolution

Comment (devin-donnelly) [inline docs/getting-started-guides/windows/README.md]: Yes, that's good.

**Comment by jbhurat** [comment]
Thanks for your review @devin-donnelly. I have made the suggested chances.

**Comment by devin-donnelly** [comment]
Awesome, @jbhurat . I'll give this a docs LGTM; I'll let @erictune handle tech LGTM, and then we can get this merged.

**Comment by erictune** [comment]
Thanks for your patience on my review.  This looks great.  Excited to see this new way to use Kubernetes and Windows!  

LGTM.
